### PR TITLE
Fix UefiVarialbeLib KW issues

### PR DESCRIPTION
**Pull request by stalamudupula**
This patch addresses KW issues reported from UefiVarialbeLib.

Signed-off-by: Sai T <sai.kiran.talamudupula@intel.com>

### CODE REVIEW

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c [hunk 1]
```diff
@@ -889,7 +892,7 @@ FindVariableEx (
       //
       // Record Variable in VariableIndex HOB
       //
-      if ((IndexTable != NULL) && !StopRecord) {
+      if ((IndexTable != NULL) && !StopRecord && (IndexTable->Length < UEFI_VARIABLE_INDEX_TABLE_VOLUME)) {
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c:895]: Do we need this code if line#845 is before for loop?

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c [hunk 2]
```diff
@@ -842,6 +842,9 @@ FindVariableEx (
     // The IndexTable->Index[Index] records the distance of two neighbouring VAR_ADDED type variables.
     //
     for (Offset = 0, Index = 0; Index < IndexTable->Length; Index++) {
+      if (Index >= UEFI_VARIABLE_INDEX_TABLE_VOLUME) {
+        return EFI_OUT_OF_RESOURCES;
+      }
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c:847]: Can we do this before for loop? I think EFI_BAD_BUFFER_SIZE would be better.
```
if (IndexTable->Length >= UEFI_VARIABLE_INDEX_TABLE_VOLUME) {
  return EFI_BAD_BUFFER_SIZE;
}
for ()
```

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c [hunk 3]
```diff
@@ -889,7 +892,7 @@ FindVariableEx (
       //
       // Record Variable in VariableIndex HOB
       //
-      if ((IndexTable != NULL) && !StopRecord) {
+      if ((IndexTable != NULL) && !StopRecord && (IndexTable->Length < UEFI_VARIABLE_INDEX_TABLE_VOLUME)) {
         Offset = (UINTN) Variable - (UINTN) LastVariable;
         if ((Offset > 0x0FFFF) || (IndexTable->Length == sizeof (IndexTable->Index) / sizeof (IndexTable->Index[0]))) {
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c:897]: This condition may resolve the issue.
`
if ((Offset > 0x0FFFF) || (IndexTable->Length == sizeof (IndexTable->Index) / sizeof (IndexTable->Index[0]))) {
`
->
`
if ((Offset > 0x0FFFF) || (IndexTable->Length >= sizeof (IndexTable->Index) / sizeof (IndexTable->Index[0]))) {
`

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c (hunk 2 above)
Comment (gdong1) [reply BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c:847]: Not sure the exact KW issue reported, but the original logic should come from EDKII https://github.com/tianocore/edk2/blob/master/MdeModulePkg/Universal/Variable/Pei/Variable.c.
Maybe check with the origin code if there is any update.

Comment (stalamudupula) [reply BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c:847]: @myguitar makes sense. Wil move.

@gdong1 No changes here in the new EDK2 driver.

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c (hunk 3 above)
Comment (stalamudupula) [reply BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c:897]: @myguitar Thanks, the new EDK2 driver already made this change, will import.

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c [hunk 4]
```diff
@@ -841,6 +841,9 @@ FindVariableEx (
     // traverse the variable index table to look for varible.
     // The IndexTable->Index[Index] records the distance of two neighbouring VAR_ADDED type variables.
     //
+    if (IndexTable->Length >= UEFI_VARIABLE_INDEX_TABLE_VOLUME) {
+      return EFI_BAD_BUFFER_SIZE;
+    }
```
[the remark this answers is not in the document]
Comment (stalamudupula) [reply BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c:846]: This is the KW critical error reported:

Array 'Index' of size 122 may use index value(s) 123..USHRT_MAX

Doesnt say which lines.

[the remark this answers is not in the document]
Comment (stalamudupula) [reply BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c:846]: KW report complains only about line # 901. will remove this check, not needed.

**Review by myguitar — APPROVED** [review]
Looks good to me

**Review by gdong1 — APPROVED** [review]

**Pull request merged**